backend/infrahub/core/relationship/model.py

```diff
@@ -725,17 +744,68 @@ def __len__(self) -> int:
 
         return len(self._relationships)
 
-    async def get_peer(self, db: InfrahubDatabase) -> Optional[Node]:
+    @overload
+    async def get_peer(
+        self,
+        db: InfrahubDatabase,
+        raise_on_error: Literal[False] = False,
+    ) -> Node | None: ...
+
+    @overload
+    async def get_peer(
+        self,
+        db: InfrahubDatabase,
+        raise_on_error: Literal[True] = True,
+    ) -> Node: ...
+
+    @overload
+    async def get_peer(
+        self,
+        db: InfrahubDatabase,
+        peer_type: PeerType,
+        raise_on_error: Literal[False] = False,
+    ) -> PeerType | None: ...
+
+    @overload
+    async def get_peer(
+        self,
+        db: InfrahubDatabase,
+        peer_type: PeerType,
+        raise_on_error: Literal[True] = True,
+    ) -> PeerType: ...
```

Comment:
It looks like the overload-overlap issue is simply due to the ordering of these overloads. I'm not completely sure if it's because you start with the default value for raise_on_error.

An updated example looks like this, here I've changed the order of the objects and also added `peer_type: Literal[None] = None` on some of the definitions.


```python
    @overload
    async def get_peer(
        self,
        db: InfrahubDatabase,
        peer_type: type[PeerType],
        raise_on_error: Literal[True] = True,
    ) -> PeerType: ...

    @overload
    async def get_peer(
        self,
        db: InfrahubDatabase,
        peer_type: type[PeerType],
        raise_on_error: Literal[False] = False,
    ) -> PeerType | None: ...

    @overload
    async def get_peer(
        self,
        db: InfrahubDatabase,
        peer_type: Literal[None] = None,
        raise_on_error: Literal[True] = True,
    ) -> Node: ...

    @overload
    async def get_peer(
        self,
        db: InfrahubDatabase,
        peer_type: Literal[None] = None,
        raise_on_error: Literal[False] = False,
    ) -> Node | None: ...
```